### PR TITLE
Fix moving channels

### LINKED ISSUE #1584: Crash when moving mixer channels to left/right
Sometimes you are able to cause a Segmentation Fault when using the "Move Right"/"Move Left" functionality on mixer channels.

I was not able to identify how to reproduce this error deterministically, but you should be able to trigger this just moving nearby channels at the rightmost part of the mixer.

The backtrace shows that this is the offending line: https://github.com/LMMS/lmms/blob/40438689922e843fcf07d22dd52378cb7e5237d5/src/gui/widgets/FxLine.cpp#L208

Full backtrace here: https://gist.github.com/badosu/54bf4c6cd2d504daaab0

### PR DESCRIPTION
**Pull request by badosu**
The FxLine widget was being deleted before the contextMenu event returned sometimes.

Please verify whether the new deletion mechanism is adequate.

I still see some weird segfaults when clicking randomly and performing random actions on the Mixer widget about `index out of range` but this is probably a separate issue.

Fixes #1584

### CODE REVIEW

File: src/gui/widgets/FxLine.cpp [hunk 1]
```diff
@@ -208,6 +208,7 @@ void FxLine::contextMenuEvent( QContextMenuEvent * )
 	contextMenu->addHelpAction();
 	contextMenu->exec( QCursor::pos() );
 	delete contextMenu;
+	delete this;
```
Comment (DanWin) [inline src/gui/widgets/FxLine.cpp]: This always causes deletion of the FxLine, even when you don't select any option in the context menu. You can even remove master with this. You should remove this line.

Comment (badosu) [inline src/gui/widgets/FxLine.cpp]: Oh yeah,  guess you're right. Thanks for catching this one.

However how can I be able to deallocate this object on the move left/right case?

File: src/gui/FxMixerView.cpp [hunk 2]
```diff
@@ -453,7 +453,6 @@ void FxMixerView::moveChannelLeft(int index)
 		delete m_fxChannelViews[i]->m_fader;
 		delete m_fxChannelViews[i]->m_muteBtn;
 		delete m_fxChannelViews[i]->m_soloBtn;
-		delete m_fxChannelViews[i]->m_fxLine;
```
Comment (badosu) [inline src/gui/FxMixerView.cpp]: @DanWin This should be the fix.

However I can't confirm that we are not leaking memory.

**Comment by badosu** [comment]
@tresf You may want to cherry pick this one to the next 1.1.1 RC

**Comment by tresf** [comment]
@badosu thanks for the hint.  It seems that there is still some undecided action on this PR, no?

**Comment by badosu** [comment]
@tresf Not really, the problem is that I removed an object deallocation without being sure if this is collected later.

I researched a bit of QT's hierarchy model to try figure out if I was safe but since the FxLine is created inside a mixer view and we never get rid of the mixer view I am not quite sure.

Of course maybe the cost of not deallocating this object is negiglible compared to crashing the application so this may be pulled in.

TL; DR: I removed a deallocation that crashed the software without moving it to another place and could not figure out where to move it.

**Comment by tresf** [comment]
@badosu Yeah, that was the impression I got.  Well, softrabbit just found an arpeggio issue with preset previews, so we now have a little more time to merge this as I don't want to release 1.1.x until after arpeggio issue fixed.

**Comment by diizy** [comment]
No, removing deallocation leads potentially to memory leaks. We can't have that.

**Comment by diizy** [comment]
Also, I don't really know why we are deleting and recreating the widget here in the first place. Looks like shoddy code... we should just be updating the widget, deleting/recreating it should not be necessary.

**Comment by tresf** [comment]
> Also, I don't really know why we are deleting and recreating the widget here in the first place. Looks like shoddy code... we should just be updating the widget, deleting/recreating it should not be necessary.

Good catch.  It looks like Andrew Kelley copied/pasted the `refreshView()` code into the `moveChannelLeft/Right()` code when he originally wrote it per https://github.com/lmms/lmms/commit/a955fb755a9b99d4a43fcbd35200d54328117d3a

**Comment by badosu** [comment]
@diizy @tresf Thanks for the feedback.

I agree that refactoring the code to preserve the widget's properties and updating it's position on the view is a much better solution, however I am a bit hesitant of touching this part right now.

I'll approach this solution as soon as I feel confident on how to implement it in a safer and better way, it may even facilitate the process of adding a handle to move the Channels drag'n'dropping like tracks in the song editor and remove this right click option altogether.

Feel free to merge or close this PR as you wish. For this particular bug in the current implementation I will let you guys decide between the lesser of two evils.

**Comment by diizy** [comment]
Not really comfortable merging something that can lead to memory leaks. Lets wait for a better solution.

**Comment by badosu** [comment]
Personally, I feel like potentially increasing a tiny amount of memory everytime a low usage part of code is accessed is much more acceptable than lettiing the software crash. But I respect your opinion and I'll push forward to devise better solutions on the future.

**Comment by tresf** [comment]
> Personally, I feel like potentially increasing a tiny amount of memory everytime a low usage part of code is accessed is much more acceptable than lettiing the software crash. 

Yeah, I tend to agree.  We're sort of "preaching to the choir" here... I look forward to a better solution.  Thanks for the understanding.

**Comment by diizy** [comment]
On 01/16/2015 10:41 PM, Tres Finocchiaro wrote:
>
>     Personally, I feel like potentially increasing a tiny amount of
>     memory everytime a low usage part of code is accessed is much more
>     acceptable than lettiing the software crash.
>
> Yeah, I tend to agree. We're sort of "preaching to the choir" here...
> I look forward to a better solution. Thanks for the understanding.
>

There's two ways to do things... do them fast, or do them right.

Ultimately, if we opt for the "fast" (or "easy") way, it'll only bite us
in the ass in the long run. It's better to do things right the first
way. LMMS has been plagued too long by shoddy code and band-aid
solutions (and I'm by no means innocent of this myself, just so you know
I'm not just pointing fingers here). It's time we change that trend.
Instead of accepting any code because someone "worked so hard" for it,
we should consider quality first. Let people learn through experience
how to solve problems, how to write good code. I myself have written
tons of code that I've eventually discarded because in practice it
didn't work that well (see: the first sampletrack-fixing attempts).

And also remember, if a problem is too hard for you to solve right now,
there's no shame in stepping back and giving someone else a shot at it -
or coming back to it later when you've learned more.

**Comment by tresf** [comment]
> There's two ways to do things... do them fast, or do them right.

> Ultimately, if we opt for the "fast" (or "easy") way, it'll only bite us
in the ass in the long run. It's better to do things right the first
way.

No one is arguing that here.  It's rather a question of if we put a patch over the boat's hole to stop a big problem from happening now, or wait for the panel to be properly repaired later.

This chunk of code was merged in and it is crashing our software.  

> there's no shame in stepping back and giving someone else a shot at it

There's also little harm here in patching the bug to prevent the crash.  The only argument against it might take our attention away from the bigger problem just to have it bite us later, but this is a question about damage control.  Hard crashes are tremendously damaging to our software, and this code is broken before and after the PR, just slightly less broken after the PR.

> Let people learn through experience how to solve problems, how to write good code.

There is the fire fighter putting out the building fire, and then there's the engineer that designs the building.  They must work together or else one of them will give up and find a new team to work with. :)

**Comment by Spekular** [comment]
Hard crash vs mem leak, tough choice. On one hand, losing all your
progress. On the other, think of all the precious RAM you're losing. You
never get that back, ever, unlike a project which is just some hours of
work.

**Comment by diizy** [comment]
On 01/17/2015 12:16 AM, Spekular wrote:
> Hard crash vs mem leak, tough choice

No it's not because that's a false dichotomy. Take the third option: do
things right.

**Comment by Spekular** [comment]
@vesa The decision being made *right now* is whether or not to merge this
pr. If we don't, we still have a hard crash. If we do, we have a mem leak
instead. My crash vs leak statement is perfectly correct.

**Comment by tresf** [comment]
@Spekular right now we ate waiting invoking option 0 (do nothing) and waiting for for option 3 (do it right). :)
